In [5]:
import numpy as np
import pandas as pd

In [22]:
from sklearn.datasets import load_boston
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from sklearn.ensemble import RandomForestRegressor

In [7]:
boston = load_boston()
data = boston.data
feature_names = boston.feature_names
X = pd.DataFrame(data, columns = feature_names)
target = boston.target
y = pd.DataFrame(target, columns = ['price'])
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.3, random_state = 42)

# Задание 1

In [27]:
lr = LinearRegression()

lr.fit(X_train, y_train)
y_pred = lr.predict(X_test)

check_test = pd.DataFrame({'y_test': y_test['price'],
                          'y_pred': y_pred.flatten()},
                         columns = ['y_test', 'y_pred'])
check_test

y_test     y_pred
173    23.6  28.648960
274    32.4  36.495014
491    13.6  15.411193
72     22.8  25.403213
452    16.1  18.855280
..      ...        ...
441    17.1  17.403672
23     14.5  13.385941
225    50.0  39.983425
433    14.3  16.682863
447    12.6  18.285618

[152 rows x 2 columns]

In [20]:
r2_score(y_test, y_pred)

0.711226005748496

# Задание 2

In [24]:
rfr = RandomForestRegressor(n_estimators= 1000, max_depth= 12, random_state= 42)

In [25]:
rfr.fit(X_train, y_train.values[:, 0])
y_pred = rfr.predict(X_test)

check_test = pd.DataFrame({'y_test': y_test['price'],
                          'y_pred': y_pred.flatten()},
                         columns = ['y_test', 'y_pred'])

In [26]:
check_test

y_test     y_pred
173    23.6  22.806412
274    32.4  31.131464
491    13.6  16.339125
72     22.8  23.810726
452    16.1  17.139521
..      ...        ...
441    17.1  13.521640
23     14.5  15.112621
225    50.0  42.808700
433    14.3  15.586103
447    12.6  16.072425

[152 rows x 2 columns]

In [28]:
rfr.score(X_train, y_train.values[:, 0])

0.9784468411299835

В задании 2 r2 получилось очень близким к 1. В то время как в первом задании - всего лишь 0.711. Однозначно, лучше работает модель RandomForestRegressor
В материалах урока я ничего не нашел о RandomForestRegressor. Поэтому, информацию мне пришлось искать здесь:
https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html

Информцию о свойстве feature_importances_ для задания 3 я получил из того же источника

# Задание 3

In [35]:
rfr_fi = rfr.feature_importances_
feature_names, rfr_fi[5], rfr_fi[12]

(array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
        'TAX', 'PTRATIO', 'B', 'LSTAT'], dtype='<U7'),
 0.4026817857034993,
 0.4158473181914483)

In [32]:
sum(rfr_fi)

1.0

Сумма всех показателей важности равна 1. Два признака - RM и LSTAT показывают наибольшую важность